codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java

```diff
@@ -95,28 +107,126 @@ public String toString() {
+
+                for (byte b : bytes) {
+                    appendEncoded(b);
+                }
+            }
         }
-        sb.append(s, 0, idx).append("%20");
-        int size = s.length();
-        idx++;
-        for (; idx < size; idx++) {
-            char c = s.charAt(idx);
-            if (c != '+') {
-                sb.append(c);
+    }
+
+    /**
+     * @see ByteBufUtil#writeUtf8(io.netty.buffer.ByteBuf, CharSequence, int, int)
+     */
+    private void encodeUtf8Component(CharSequence s) {
+        for (int i = 0; i < s.length(); i++) {
+            char c = s.charAt(i);
+            if (c < 0x80) {
+                if (dontNeedEncoding(c)) {
+                    uriBuilder.append(c);
+                } else {
+                    appendEncoded(c);
+                }
+            } else if (c < 0x800) {
+                appendEncoded(0xc0 | (c >> 6));
+                appendEncoded(0x80 | (c & 0x3f));
+            } else if (StringUtil.isSurrogate(c)) {
+                if (!Character.isHighSurrogate(c)) {
+                    appendEncoded(WRITE_UTF_UNKNOWN);
+                    continue;
+                }
+                // Surrogate Pair consumes 2 characters.
+                if (++i == s.length()) {
+                    appendEncoded(WRITE_UTF_UNKNOWN);
+                    break;
+                }
+                // Extra method to allow inlining the rest of writeUtf8 which is the most likely code path.
+                writeUtf8Surrogate(c, s.charAt(i));
             } else {
-                sb.append("%20");
+                appendEncoded(0xe0 | (c >> 12));
+                appendEncoded(0x80 | ((c >> 6) & 0x3f));
+                appendEncoded(0x80 | (c & 0x3f));
             }
         }
     }
+
+    private void writeUtf8Surrogate(char c, char c2) {
+        if (!Character.isLowSurrogate(c2)) {
+            appendEncoded(WRITE_UTF_UNKNOWN);
+            appendEncoded(Character.isHighSurrogate(c2) ? WRITE_UTF_UNKNOWN : c2);
+            return;
+        }
+        int codePoint = Character.toCodePoint(c, c2);
+        // See http://www.unicode.org/versions/Unicode7.0.0/ch03.pdf#G2630.
+        appendEncoded(0xf0 | (codePoint >> 18));
+        appendEncoded(0x80 | ((codePoint >> 12) & 0x3f));
+        appendEncoded(0x80 | ((codePoint >> 6) & 0x3f));
+        appendEncoded(0x80 | (codePoint & 0x3f));
+    }
+
+    private void appendEncoded(int b) {
+        uriBuilder.append('%').append(forDigit(b >> 4)).append(forDigit(b));
+    }
+
+    /**
+     * Convert the given digit to a upper hexadecimal char.
+     * <p>
+     * the 55 in this method means {@code 'A' - 10}
+     *
+     * @param digit the number to convert to a character.
+     * @return the {@code char} representation of the specified digit
+     * in hexadecimal.
+     */
+    private static char forDigit(int digit) {
+        digit = digit & 0xF;
+        return (char) (digit < 10 ? '0' + digit : 55 + digit);
```

Comment:
Could be made branchless?

Reply:
I am sorry...I think about it for almost a hour and I don‘t know how to make this branchless...If the branchless means remove the use of `?` or `if`

Reply:
you mean this?
```
private static final char[] CHAR_MAP = {'0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'};
private static char forDigit(int digit) {
        ...
        return CHAR_MAP[digit & 0xF];
       ...
}
```

Reply:
you can try this:
```
      digit = digit & 0xF;
      final int isLessOrEqual9 = (digit - 10) >> 31;
      final int bin2hexAsciiDistance = 48 + ((~isLessOrEqual9)&7);
      return (char) (digit + bin2hexAsciiDistance);
```

Reply:
I write a simple benchmark for this
```
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@Measurement(iterations = 5,batchSize = 10000000)
@Fork(1)
@Warmup(iterations = 3)
@Threads(1)
public class DigitConvertBenchmark {

    private static final char[] CHAR_MAP = {'0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'};

    @Benchmark
    public int baseLine() {
        return ThreadLocalRandom.current().nextInt(0,16);
    }

    @Benchmark
    public char forDigit() {
        int digit = ThreadLocalRandom.current().nextInt(0,16);
        digit = digit & 0xF;
        return (char) (digit < 10 ? '0' + digit : 55 + digit);
    }

    @Benchmark
    public char forDigitBranchLess() {
        int digit = ThreadLocalRandom.current().nextInt(0,16);
        digit = digit & 0xF;
        final int isLessOrEqual9 = (digit - 10) >> 31;
        final int bin2hexAsciiDistance = 48 + ((~isLessOrEqual9)&7);
        return (char) (digit + bin2hexAsciiDistance);
    }

    @Benchmark
    public char forDigitArray() {
        int digit = ThreadLocalRandom.current().nextInt(0,16);
        digit = digit & 0xF;
        return CHAR_MAP[digit];
    }
}
```

and the result :

```
Benchmark                                 Mode  Cnt   Score   Error  Units
DigitConvertBenchmark.baseLine            avgt    5  24.598 ± 0.126  ms/op
DigitConvertBenchmark.forDigit            avgt    5  58.897 ± 0.593  ms/op
DigitConvertBenchmark.forDigitArray       avgt    5  25.697 ± 0.377  ms/op
DigitConvertBenchmark.forDigitBranchLess  avgt    5  28.141 ± 1.008  ms/op
```

The branchless approach is actually more effective, If the benchmark code was right...And which way should we take？The array one or the logic one?

Reply:
Try not use batchSize and increase @Fork to 10 at least :)
And switch to smaller OutputTimeUnit

Reply:
> Try not use batchSize and increase @fork to 10 at least :)
> And switch to smaller OutputTimeUnit

use `for loop` or just run a single time?

Reply:
I'm running a similar bench to understand what's going on :+1:

Reply:
My got feeling tell that although the array approach seems to faster, in a real application code, the branchless is to be preferred (I know is complex the code): in real prod code we have cache trashing, while the branchless approach is purely CPU intensive :)
I'm trying to reproduce it with a benchmark, but is not easy :)

Reply:
maybe lets just use the array one for now...

Reply:
This one is not reproducing the cache trashing but should be somehow better, I just don't like to use the thread local random:
```
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Measurement(iterations = 5, time = 1)
@Fork(3)
@Warmup(iterations = 5, time = 2)
@Threads(1)
@State(Scope.Benchmark)
public class DigitConvertBenchmark {

    @Param({"10"})
    private int size;
    private int origin;
    private int bound;
    private static final char[] CHAR_MAP = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};

    @Setup
    public void init() {
        origin = 0;
        bound = 16;
    }

    @Benchmark
    public int baseLine() {
        return nextDigit();
    }

    private int nextDigit() {
        return ThreadLocalRandom.current().nextInt(origin, bound);
    }

    @Benchmark
    public int forDigit() {
        int total = 0;
        for (int i = 0, size = this.size; i < size; i++) {
            int digit = nextDigit();
            digit = digit & 0xF;
            total += ((char) (digit < 10 ? '0' + digit : 55 + digit));
        }
        return total;
    }

    @Benchmark
    public int forDigitBranchLess() {
        int total = 0;
        for (int i = 0, size = this.size; i < size; i++) {
            int digit = nextDigit();
            digit = digit & 0xF;
            final int isLessOrEqual9 = (digit - 10) >> 31;
            final int bin2hexAsciiDistance = 48 + ((~isLessOrEqual9) & 7);
            total += ((char) (digit + bin2hexAsciiDistance));
        }
        return total;
    }

    @Benchmark
    public int forDigitArray() {
        int total = 0;
        for (int i = 0, size = this.size; i < size; i++) {
            int digit = nextDigit();
            digit = digit & 0xF;
            total += CHAR_MAP[digit];
        }
        return total;
    }
}
```

Reply:
@Liyixin95 I'm getting results similar to yours, so do as @normanmaurer suggest too :) 
Actually the array approach seems faster on benchmarks and the code is simpler, just wanna save cache lines for other stuff, that's why I've suggested the other approach, but I'm happy anyway :+1:

Reply:
> @Liyixin95 I'm getting results similar to yours, so do as @normanmaurer suggest too :)
> Actually the array approach seems faster on benchmarks and the code is simpler, just wanna save cache lines for other stuff, that's why I've suggested the other approach, but I'm happy anyway 👍

All right,I will use that array approach, and thanks for theaching me a lot.

@normanmaurer  Besides,should I add the array and the `forDigit ` method to another UtilsClass?For sharing the array with other function.

Reply:
I think we do not need a utility class for now.